src/backend/distributed/commands/multi_copy.c

```diff
@@ -0,0 +1,1177 @@
+	columnCount = tupleDescriptor->natts;
+	columnValues = palloc0(columnCount * sizeof(Datum));
+	columnNulls = palloc0(columnCount * sizeof(bool));
+
+	/* load the list of shards and verify that we have shards to copy into */
+	shardIntervalList = LoadShardIntervalList(tableId);
+	if (shardIntervalList == NIL)
+	{
+		if (partitionMethod == DISTRIBUTE_BY_HASH)
+		{
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("could not find any shards for query"),
+							errdetail("No shards exist for distributed table \"%s\".",
+									  relationName),
+							errhint("Run master_create_worker_shards to create shards "
+									"and try again.")));
+		}
+		else
+		{
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("could not find any shards for query"),
+							errdetail("No shards exist for distributed table \"%s\".",
+									  relationName)));
+		}
+	}
+
+	/* create a mapping of shard id to a connection for each of its placements */
+	shardConnectionHash = CreateShardConnectionHash();
+
+	/* lock shards in order of shard id to prevent deadlock */
+	shardIntervalList = SortList(shardIntervalList, CompareShardIntervalsById);
+
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		int64 shardId = shardInterval->shardId;
+
+		/* prevent concurrent changes to number of placements */
+		LockShardDistributionMetadata(shardId, ShareLock);
+
+		/* prevent concurrent update/delete statements */
+		LockShardResource(shardId, ShareLock);
+	}
+
+	/* initialize the shard interval cache */
+	shardCount = list_length(shardIntervalList);
+	shardIntervalCache = SortedShardIntervalArray(shardIntervalList);
+
+	/* determine whether to use binary search */
+	if (partitionMethod != DISTRIBUTE_BY_HASH ||
+		!IsUniformHashDistribution(shardIntervalCache, shardCount))
+	{
+		useBinarySearch = true;
+	}
+
+	/* initialize copy state to read from COPY data source */
+	copyState = BeginCopyFrom(rel, copyStatement->filename,
+							  copyStatement->is_program,
+							  copyStatement->attlist,
+							  copyStatement->options);
+
+	/* set up callback to identify error line number */
+	errorCallback.callback = CopyFromErrorCallback;
+	errorCallback.arg = (void *) copyState;
+	errorCallback.previous = error_context_stack;
+	error_context_stack = &errorCallback;
+
+	executorState = CreateExecutorState();
+	executorTupleContext = GetPerTupleMemoryContext(executorState);
+	executorExpressionContext = GetPerTupleExprContext(executorState);
+
+	copyOutState = (CopyOutState) palloc0(sizeof(CopyOutStateData));
+	copyOutState->binary = true;
+	copyOutState->fe_msgbuf = makeStringInfo();
+	copyOutState->rowcontext = executorTupleContext;
+
+	columnOutputFunctions = ColumnOutputFunctions(tupleDescriptor, copyOutState->binary);
+
+	/* we use a PG_TRY block to roll back on errors (e.g. in NextCopyFrom) */
+	PG_TRY();
+	{
```

Comment:
On 2016-03-25 08:51:14 -0700, Marco Slot wrote:
> I'd rather not put anything in the PG_TRY block that doesn't need to be there, since the error handling requires more thought.

It's actually the other way round imo. Anything that can be touched both
inside and outside the PG_TRY has to be thought about much more
careful. PG_TRY/CATCH is implemented using siglongjmp.  Quoting man
sigsetjmp:

       The compiler may optimize variables into registers, and longjmp() may restore the values of other registers  in  addition  to
       the  stack  pointer  and  program  counter.   Consequently, the values of automatic variables are unspecified after a call to
       longjmp() if they meet all the following criteria:

       ·  they are local to the function that made the corresponding setjmp(3) call;

       ·  their values are changed between the calls to setjmp(3) and longjmp(); and

       ·  they are not declared as volatile.